File: my-folder/0017-letter-combinations-of-a-phone-number/solution.java
class Solution {
    public List<String> letterCombinations(String digits) {
        if (digits == null || digits.isEmpty()) return new ArrayList<>();
        String[] mappings = {"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        List<String> ans = new ArrayList<>();
        ans.add("");

        for (char digit : digits.toCharArray()) {
            List<String> extendCombination = new ArrayList<>();
            for (String currentCombination : ans) {
                for (char newChar : mappings[digit - '2'].toCharArray()) {
                    extendCombination.add(currentCombination + newChar);
                }
            }
            ans = extendCombination;
        }

        return ans;     
    }
}

File: my-folder/0017-letter-combinations-of-a-phone-number/solution.cpp
class Solution {
public:
    void backtracking(vector<string>& mapping, vector<string>& res, string tmp, int index, string digits) {
        if(index == digits.length()) {
            res.push_back(tmp);
            return;
        }
        for(char c : mapping[digits[index]-'2']) {
            tmp += c;
            backtracking(mapping, res, tmp, index+1, digits);
            tmp.pop_back();
        }    
    }
    vector<string> letterCombinations(string digits) {
        if(digits.empty()){
            return {};
        }
        vector<string> mappings{"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"}, ans;
        backtracking(mappings, ans, "", 0, digits);
        return ans;
    }
};
